### ISSUE TITLE
Is narrow attention implemented correctly?

### ISSUE THREAD

**Issue by TheGrayFrost** [pbloem/former#13]
Hey man! Thanks so much for writing this amazing exposition on transformers. I understand things so much better because of you. And the [einsum link](https://rockt.github.io/2018/04/30/einsum) just changed my life for ever.

I had a doubt regarding the Narrow Self-Attention implementation...

In `former/modules.py`, class `SelfAttentionNarrow` you have the following code for `__init__`:
https://github.com/pbloem/former/blob/6a3295c94a151291de6838b98826da13e5f57bf3/former/modules.py#L90-L95

Now in the `forward` method you do:
https://github.com/pbloem/former/blob/6a3295c94a151291de6838b98826da13e5f57bf3/former/modules.py#L105-L110

Does this not mean that you only have one attention head rather than `h` of them? As all the `h` `s`-sized pieces of `x` are going through the same linear layer.
From my understanding there should be `h` Linear layers of size `s` x `s` each. Or a single weight matrix of shape `h` x `s` x `s` depending on your implementation. 
Am I going wrong somewhere? It would be really awesome if you clarify.

**Comment by TheGrayFrost** [comment]
I saw that in the Annotated Transformer code, they make the linear layer of size `e`x`e`. So after the linear layer you get shape `b`x`t`x`e`, which they reshape to `b`x`t`x`h`x`s`. And then, they treat all of these Q, K, V vectors as having the `h` heads each of size `s`.
So, in their code, it's not that each head is looking at a `s`-sized chunk of `x`. All the "heads" looks at all of `x`, and sort of produce `h` `s`-sized (Q, K, V) vectors. 

So,  it's like wide attention only, just that the dimension output of each head size has reduced from `e` to `s`. But we get the advantage that:
1. the linear layer size goes down from `h`x`e`x`e` to just `e`x`e` (and the compute advantage from that size reduction)
2. the final `h * e`x`e` sized `unifyheads` is not needed (and the compute advantage from not having to do unify heads)

Maybe it all comes down to implementation. You can do whatever way you want. It's just affecting performance. 
But I am still confused if we can consider your implementation to have `h` heads.

**Comment by pbloem** [comment]
I think you're right... this is just plain wrong (not sure what I was thinking when I wrote this). I'll have a closer look at how the other implementations do this, and update the code.

**Comment by pbloem** [comment]
Actually, it's not quite as wrong as I thought. Each head is performing the same transformation, but they're looking at separate chunks of the input vector. So the differences in behavior between the heads are purely determined by what's in the chunks, but they are also completely different from head to head. 

I'm not sure how much power you'd actually add by giving each head it's own transformation as well. Still, I'll find some time to dig into the other implementations.

**Comment by TheGrayFrost** [comment]
I am not sure if you could consider them separate heads if they don't have separate parameters. I mean, shouldn't a head  denote _who_ is doing the transformation rather than _what_ is being transformed?
Like, in wide attention also, all the heads are transforming the same `x`, but that is fine as each head is doing its own thing.

**Comment by TheGrayFrost** [comment]
Sorry to keep disturbing you like this...
This is unrelated to the narrow attention thing.
Actually, there is this one line in the blog that was bugging me. I don't want to create an issue just for it but I have to ask:

<img src="https://user-images.githubusercontent.com/24153594/85807510-c1dd9a80-b76f-11ea-8fc5-b721f9f4dc4a.png"  width=400>

I think if you change "Mary gave Roses to Susan" to "Susan gave Roses to Mary", multi-headed attention still won't change the vector for `gave` - as self-attention layer is completely position independent. This permutation issue is addressed by positional encodings/embeddings. I understood that with each head we want to capture different contexts, but this example doesn't seem to represent that motivation.

I really don't want to sound like I am nitpicking. It's the most amazing blog, and I just thought this might confuse someone.

**Comment by shabie** [comment]
> Sorry to keep disturbing you like this...
> This is unrelated to the narrow attention thing.
> Actually, there is this one line in the blog that was bugging me. I don't want to create an issue just for it but I have to ask:
> 
> <img alt="" width="400" src="https://user-images.githubusercontent.com/24153594/85807510-c1dd9a80-b76f-11ea-8fc5-b721f9f4dc4a.png">
> 
> I think if you change "Mary gave roses to Susan" to "Susan gave roses to Mary", multi-headed attention still won't change the vector for `gave` - as self-attention layer is completely position independent. This permutation issue is addressed by positional encodings/embeddings. I understood that with each head we want to capture different contexts, but this example doesn't seem to represent that motivation.
> 
> I really don't want to sound like I am nitpicking. It's the most amazing blog, and I just thought this might confuse someone.

I was thinking exactly the same. Otherwise positional embeddings are redundant.

I mean in narrow multi-head attention the answer seems to be that it pays attention to subspaces in the vector representation of the word (since one token's vector is chopped up into chunks).

In the broad multi-head attention, by extension, I guess the logic remains the same. This is somewhat similar to how the convolution kernels learn different base features despite convolving over the same image.

It would be nice to get some clarification on this since this blog post on transformers by @pbloem is breathtakingly well-written.

**Comment by SubhashPadala** [comment]
**Why √k? Imagine a vector in ℝ^k with values all c. Its Euclidean length is √kc. Therefore, we are dividing out the amount by which the increase in dimension increases the length of the average vectors.**

In SelfAttentionNarrow, we are increasing our space by s=emb/heads, so our space will increase by a factor of √s, but why are we dividing by √k ?

**Comment by pbloem** [comment]
The code now contains a canonical self-attention that is as far as I can tell precisely what is used in Vaswani and all following work. I've also added a stripped down version of the self attention as implemented in GPT2 (the huggingface port). These behave almost exactly the same.

I've checked all four versions  (canonical, gpt2, narrow, wide) on a simple experiment and the narrow option shown above is considerably worse. Interestingly, the wide version is much slower and memory hungry (as expected), but for a fixed time budget seems to reach a similar performance to the canonical one (i.e. it gets better performance in fewer batches).
